Prevent node from running and installing on Windows Vista or earlier

orangemocha opened the pull request:
Ref: https://github.com/nodejs/node/issues/3804
R: @nodejs/platform-windows @ChALkeR 

The last commit is not necessary to achieve the objective, but as a consequence of supporting Windows 7 or greater we can use features from Windows Installer 5.0.

Tested on Vista and Win7 (boundary versions). Screenshots of outcomes on Vista:

Running node.exe
![windows vista-2016-02-09-20-07-55 - copy](https://cloud.githubusercontent.com/assets/1625870/12929139/a503bb84-cf47-11e5-968c-0632ea77179e.png)

Launching the msi, with Windows Installer 5.0 (it's available as an update)
![windows vista-2016-02-09-20-13-03 - copy](https://cloud.githubusercontent.com/assets/1625870/12929188/d3afa786-cf47-11e5-8c2e-7f23c60152bd.png)

Launching the msi, with Windows Installer < 5.0
![windows vista-2016-02-09-20-42-19 - copy](https://cloud.githubusercontent.com/assets/1625870/12929206/e463eb46-cf47-11e5-9763-7537816dce09.png)

**UPDATE**
The latest version of this change only shows the second dialog during MSI install.
When running node.exe, an error is printed on stderr.

ChALkeR commented:
@orangemocha Yay, thanks for this and for mentioning me. I can not test it though, because I don't have a Vista or XP setup.

From your screenshots, the second message is clear, but the first and the last ones are not. Are you sure that the remaining users of XP/Vista won't come here with those unobvious error messages? Is there any way to make it clear that their platform is deprecated and that it's not a random bug?

orangemocha commented:
@ChALkeR good points. I am afraid that most users of the installer will see the third error, which is indeed misleading. Unfortunately if we are going to upgrade to Windows Installer 5.0 there's no way that the MSI can run on the unsupported platforms, so there's now way to give a better error. In light of that, it might be better to leave the last commit out. @joaocgreis , do you know of any critical or important features that we need from Windows Installer 5.0?

The first error would only be presented to people who download node.exe instead of using the installer. I am not sure what percentage of users do that. As an alternative to that change, we could maintain the subsystem compatibility and add a runtime check in main(), to simply print an error message and exit on unsupported versions.

rvagg commented on tools/msvs/msi/product.wxs:
```diff
@@ -19,10 +19,14 @@
              InstallScope="perMachine"/>
 
+    <Condition Message="This application is only supported on Windows 7, Windows Server 2008 R2, or higher.">
+      <![CDATA[Installed OR (VersionNT >= 601)]]>
```
This works as a negative condition? if this is false then the message is displayed?

Also, does `Installed` here mean that if you already have Node installed and try to install v6 then it's not going to give you this error?

joaocgreis replied:
The idea here is that this is placed in a condition table in the MSI. The installer will only launch if the conditions are all true.

`Installed` is true when uninstalling and false when installing. @orangemocha we must make sure this works as intended when upgrading, in both Vista and 7.

orangemocha replied:
> This works as a negative condition? if this is false then the message is displayed?

Correct. The condition must be true for the installer to continue, otherwise the message is displayed.

I have tested a variety of upgrade scenarios and the current condition works fine. As noted, `Installed` is true when uninstalling and false when installing, *and it refers to the current MSI, not any other installed versions of the product*. So the `Installed` property in the condition is to make sure that this version of the MSI with the condition can always be repaired or removed, even if the rest of the condition became false. The OS version could change due to an OS upgrade, or a service pack removal. This probably is not possible with Windows 7, but it looks like a solid pattern to follow (and it's the recommended way in [Wix documentation](http://wixtoolset.org/documentation/manual/v3/howtos/redistributables_and_install_checks/block_install_on_os.html)).

rvagg commented:
lgtm pending testing of the upgrade operation as per comments above

joaocgreis commented:
Regarding `InstallerVersion="500"`, that's a good point, we should leave it out to display the clear error.

This is mostly a nice to have (re #4111, and our custom actions could perhaps be simplified), and this seemed the proper place to do it. This will not force that exact version of the installer, it would just block installing with older versions. Actually, #4111 might be clear to land with or without this.

orangemocha commented:
I have reworked this PR based on the feedback. I have removed the upgrade to Windows Installer 5.0, deemed unnecessary, so the error message given by the installer is clear.
I have also modified the way that node.exe is prevented from running. Instead of making it incompatible with the subsystem version, I made it simply made it check the version at runtime, and print a message and exit if the required version is not met. PTAL

orangemocha commented:
.. the runtime error should also make it easier for any motivated XP/Vista users to hack their way around it, and at that point they should understand that they are on their own.

orangemocha commented:
Ping @ChALkeR or @nodejs/platform-windows for review. Thanks!

joaocgreis commented:
LGTM

Tested, including upgrading from 5.7 on XP.

orangemocha commented:
Landed in 55db19074dc35be99a610009f9fe9bb27ec4a5c6 and 1cf26c036daf438ea50f0d7aa4984f8d67fa792b. Thank you!
